src/src/shared/definitions.rs

```diff
@@ -0,0 +1,94 @@
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct TableId(pub String);
+
+#[derive(Clone, PartialEq, Eq, Hash, Debug)]
+pub struct PageId(pub usize);
+
+pub mod page {
+    use super::*;
+
+    #[derive(Debug)]
+    pub enum PageData<'a> {
+        Safe(&'a [u8]),
+        Unsafe(*const u8),
+    }
+
+    #[derive(Debug)]
+    pub struct Page<'a> {
+        pub id: PageId,
+        pub size: usize,
+        data: PageData<'a>,
+    }
+
+    impl<'a> Page<'a> {
+        pub fn new_safe(id: PageId, size: usize, data: &'a [u8]) -> Self {
+            assert!(size > 0);
+            Self {
+                id,
+                size,
+                data: PageData::Safe(data),
+            }
+        }
+
+        /// # Safety
+        /// The number of bytes in the array pointed to by `data` must be equal to `size`.
+        pub unsafe fn new_unsafe(id: PageId, size: usize, data: *const u8) -> Self {
+            assert!(size > 0);
+            assert!(!data.is_null());
+            Self {
+                id,
+                size,
+                data: PageData::Unsafe(data),
+            }
+        }
+
+        pub fn is_unsafe(&self) -> bool {
+            matches!(self.data, PageData::Unsafe(_))
+        }
+
+        /// # Safety
+        /// The value of `index` must be less than `self.size`.
+        pub fn index(&self, index: usize) -> Option<u8> {
+            if index >= self.size {
+                None
+            } else {
+                match self.data {
+                    PageData::Safe(safe) => Some(safe[index]),
+                    PageData::Unsafe(r#unsafe) => {
+                        Some(unsafe { *r#unsafe.offset(index.try_into().ok()?).as_ref().unwrap() })
+                    }
+                }
+            }
+        }
+
+        pub fn iter(&'a self) -> PageIterator<'a> {
+            PageIterator::from_beginning(self)
+        }
+    }
+
+    #[derive(Debug)]
+    pub struct PageIterator<'a> {
+        current: usize,
+        page: &'a Page<'a>,
+    }
```

Comment:
When would we need to iterate over pages in buffer pool? What order would it access the pages in?

If it's for sequential access, we should take advantage of read_ahead and we should be seeing calls to the storage handler as well?

Reply:
This isn't for iterating over pages, it's an Iterator over the bytes in a single page.

Reply:
Do we need to do that?